keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,360 @@
+
+
+### Failure recovery
+
+As stated in the goals section, the migration has to make progress even if the
+environment is flaky. This section describes how the migrator recovers from
+failure.
+
+Kubernetes **replicaset controller** restarts the **migration controller** `pod`
+if it fails. The restarted **migration controller** follows the same bootstrap
+workflow as described in the [workflow section][]. Because the **migration
+controller** adopts existing `migration` objects, it can resume from where
+it left off.
+
+Kubernetes **job controller** ensures the completion of a **migration worker**.
+The **migration worker** uses the continue token as the checkpoint.
+* the worker stores the latest continue token returned by the apiserver in a
+  `configmap`.
+* when the worker starts, it resumes listing with the continue token.
+* the **migration controller** creates the configmap when creates the worker `job`,
+  and sets the worker job as the owner. When the **migration controller** deletes
+  the completed `job`, the `configmap` is garbage collected.
+
+[workflow section]:#alpha-workflow
+
+### Beta workflow - Automation
+
+It is a beta goal to automate the migration workflow. That is, migration does
+not need to be triggered manually by cluster admins, or by custom control loops
+of Kubernetes distributions.
+
+The automated migration should work for Kubernetes built-in resource types,
+custom resources, and aggregated resources.
+
+Here are two alternatives and their requirements.
+
+#### a. Exposing the configured storage versions via API
+
+This is what the [Active Management][] design proposes. It requires:
+* exposing the configured [storage versions][] of Kubernetes built-in resource
+  types and aggregated resources. 
+
+Note that custom resources already exposes this information. 
+
+The migration of built-in resource types, aggregated resources, and custom
+resources can then all be triggered by changes to the exposed storage versions
+configurations.
+
+[Active Management]:https://docs.google.com/document/d/1eoS1K40HLMl4zUyw5pnC05dEF3mzFLp5TPEEt4PFvsM
+[storage versions]:https://github.com/kubernetes/kubernetes/blob/db9545e69e630e5bc8c3c2dab14f9339dd3b871c/pkg/kubeapiserver/options/storage_versions.go#L98
+
+#### b. Using the apiserver version to trigger migration
+
+If the Kubernetes deprecation [policy][] is followed by every apiserver, and
+every cluster uses the default [storage version configs][], then we can use the
+apiserver version to drive the migration. More specifically,
+* kube-apiserver removes the `--storagve-version` flag, so that every deployment
+  of kube-apiserver uses the default storage versions. [#67678][]
+* aggregated apiservers follow semantic versioning, and expose their version via
+  API.
+* aggregated apiservers follow the same API deprecation [policy][].
+
+Then the migration of built-in resources and aggregated resources can be driven
+by the changes of apiservers versions.
```

Comment:
I'm pretty against driving storage changes on apiserver binary version changes. Many times when it changes there's no storage change needed.

Reply:
> Many times when it changes there's no storage change needed.

Migrating objects whose storage version hasn't changed doesn't impact correctness. The extra workload is also tolerable.

If storage versions can change when apiserver binary versions do *not* change, that will impact correctness. Can you suggest cases where we need to do that?

Reply:
> I'm pretty against driving storage changes on apiserver binary version changes. Many times when it changes there's no storage change needed.

Given the fairly low cost of migration and the fairly low frequency of binary versions, it seems reasonable to kick the storage migration globally when it happens.  As @caesarxuchao noted, it doesn't impact correctness and it can be refined and extended as time goes on.

As @liggitt has noted in the past, the leaking the storage version isn't required by clients and doesn't mean anything to every other client out there.

Choosing to trigger on this now doesn't preclude triggering on other items later on.  For a beta, this seems like a very simple start that doesn't require exposing extraneous information, doesn't limit future choice, and doesn't impact correctness.

Reply:
I view the kubernetes version and the api machinery version as orthogonal things. The apiserver binary version isn't logically correlated to the api versions. I would like to live in a world where I can keep my kubernetes version constant while picking up new api machinery changes, or vice versa.

Today these things happen to be correlated but that's accidental. It's really bad to make an automatic trigger that adds an unintended semantic to some facet of our api. E.g. is the /version endpoint even a "real" api covered by our deprecation policy? Adding this trigger changes the meaning of a version change from "I'm now running X" to "please touch every object in the system". I don't think that's a good meaning change. You should have to do something more than push a new patch to get every object in your system touched.

Just to be extra clear, my objection has nothing to do with efficiency and everything to do with not adding unintended consequences. 

I don't care about the exact mechanism we use as long as it has no false positives and no false negatives, again not because of efficiency but because that indicates it's being driven from a signal that's not going to change meaning in the future. We don't HAVE to expose the storage version, although that'd be a convenient thing to trigger on--it works great for rolling backwards as well as rolling forwards. Other clients don't need to know the storage version, but I think the storage migration triggerer component does deserve to know.

Are you worried that clients will abuse knowing the storage version? How?

I also really want some way of tracing resources when they move between groups, e.g. deployment. We haven't done too much of that recently, but I bet CRD authors are going to want to do it sooner or later.